community/kernel/src/main/java/org/neo4j/kernel/impl/api/SafeFreeIdManagement.java

```diff
@@ -0,0 +1,166 @@
+{
+    private final NeoStores neoStores;
+    private final JobScheduler jobScheduler;
+    @SuppressWarnings( "unchecked" )
+    private final DelayedBuffer<KernelTransactions.Snapshot>[/*index is IdType#ordinal()*/] buffers =
+            new DelayedBuffer[IdType.values().length];
+    private JobHandle maintenanceJob;
+
+    public SafeFreeIdManagement( NeoStores neoStores, JobScheduler jobScheduler )
+    {
+        this.neoStores = neoStores;
+        this.jobScheduler = jobScheduler;
+    }
+
+    // This oddity is here to break this circle:
+    // KernelTransactions -> TransactionCommitProcess -> TransactionRepresentationStoreApplier -> SafeFreeIdManagement
+    //         ^-----------------------------------------------------------------------------------------------|
+    public void initialize( Supplier<KernelTransactions.Snapshot> transactions )
+    {
+        buffers[IdType.PROPERTY.ordinal()] = buffer( neoStores.getPropertyStore(), transactions );
+        buffers[IdType.ARRAY_BLOCK.ordinal()] =
+                buffer( neoStores.getPropertyStore().getArrayStore(), transactions );
+        buffers[IdType.STRING_BLOCK.ordinal()] =
+                buffer( neoStores.getPropertyStore().getStringStore(), transactions );
+        buffers[IdType.NODE_LABELS.ordinal()] =
+                buffer( neoStores.getNodeStore().getDynamicLabelStore(), transactions );
+
+        // A little check to see that we instantiated all relevant buffers
+        for ( IdType idType : IdType.values() )
+        {
+            if ( idType.allowAggressiveReuse() && buffers[idType.ordinal()] == null )
+            {
+                throw new AssertionError( "Some buffers were forgotten, for example " + idType );
+            }
+        }
+    }
+
+    private DelayedBuffer<KernelTransactions.Snapshot> buffer( final CommonAbstractStore store,
+            final Supplier<KernelTransactions.Snapshot> transactions )
+    {
+        return new DelayedBuffer<>(
+                new Supplier<KernelTransactions.Snapshot>()
+                {
+                    @Override
+                    public KernelTransactions.Snapshot get()
+                    {
+                        return transactions.get();
+                    }
+                }, new Predicate<KernelTransactions.Snapshot>()
+                {
+                    @Override
+                    public boolean test( KernelTransactions.Snapshot snapshot )
+                    {
+                        return snapshot.allHaveClosed();
+                    }
+                }, 10_000, new Consumer<long[]>()
+                {
+                    @Override
+                    public void accept( long[] freedIds )
+                    {
+                        for ( long id : freedIds )
+                        {
+                            store.freeId( id );
+                        }
+                    }
+                } );
```

Comment:
That's a lot of anonymous inner classes. If this class got moved to `store.id` alongside `DelayedBuffer`, then `DelayedBuffer` could become package private, and depend on the `SafeFreeIdManagement` class directly.